experimental/dds/tree2/src/core/rebase/utils.ts

```diff
@@ -4,6 +4,199 @@
  */
 
 import { assert } from "@fluidframework/common-utils";
+import { ChangeRebaser, TaggedChange, tagRollbackInverse } from "./changeRebaser";
+import { GraphCommit, mintRevisionTag } from "./types";
+
+/**
+ * Rebases a source branch onto another commit in a target branch.
+ *
+ * A "branch" is defined as a "head" commit and all ancestors of that commit, i.e. one linked list in a graph of commits.
+ *
+ * The source and target branch must share an ancestor.
+ * @param changeRebaser - the change rebaser responsible for rebasing the changes in the commits of each branch
+ * @param sourceHead - the head of the source branch, which will be rebased onto `targetHead`
+ * @param targetHead - the commit to rebase the source branch onto
+ * @returns the head of a rebased source branch and the cumulative change to the source branch
+ * @remarks While a single branch must not have multiple commits with the same revision tag (that will result in undefined
+ * behavior), there may be a commit on the source branch with the same revision tag as a commit on the target branch. If such
+ * a pair is encountered while rebasing, it will be "cancelled out" in the new branch. For example:
+ * ```
+ * // (A)-(B)-(C) <- Branch X
+ * //   \
+ * //   (B')-(D) <- Branch Y
+ * //
+ * // As Branch Y is rebased onto Branch X, commits B and B' cancel out so there is no version of B on the new rebased source branch
+ * //
+ * // (A)-(B)-(C) <- Branch X
+ * //           \
+ * //           (D') <- Branch Y'
+ * //
+ * ```
+ */
+export function rebaseBranch<TChange>(
+	changeRebaser: ChangeRebaser<TChange>,
+	sourceHead: GraphCommit<TChange>,
+	targetHead: GraphCommit<TChange>,
+): [newSource: GraphCommit<TChange>, sourceChange: TChange];
+
+/**
+ * Rebases a source branch onto another commit in a target branch.
+ *
+ * A "branch" is defined as a "head" commit and all ancestors of that commit, i.e. one linked list in a graph of commits.
+ *
+ * The source and target branch must share an ancestor.
+ * @param changeRebaser - the change rebaser responsible for rebasing the changes in the commits of each branch
+ * @param sourceHead - the head of the source branch, which will be rebased onto `newBase`
+ * @param newBase - the commit to rebase the source branch onto.
+ * @param targetHead - The head of the branch that `newBase` belongs to. Must be `newBase` or a descendent of `newBase`.
+ * @returns the head of a rebased source branch and the cumulative change to the source branch
+ * @remarks This function will rebase the source branch _farther_ than `newBase` if the source branch's next commits after
+ * `newBase` match those on the target branch. For example:
+ * ```
+ * // (A)-(B)-(C)-(D)-(E) <- Branch X
+ * //   \
+ * //   (B')-(D')-(F) <- Branch Y
+ * //
+ * // If Branch Y is rebased onto commit C of Branch X, the branches must at least look like this afterwards:
+ * //
+ * // (A)-(B)-(C)-(D)-(E) <- Branch X
+ * //           \
+ * //           (B'')-(D'')-(F') <- Branch Y'
```

Comment:
This seems wrong. Shouldn't B'' be removed in this scenario?

Reply:
Oh yea good catch. Branch Y is in an invalid state because it has two versions of B in its ancestry.